python/python/lance/indices.py

```diff
@@ -0,0 +1,253 @@
+        """
+        The number of partitions / centroids in the IVF model
+        """
+        return len(self.centroids)
+
+    def save(self, uri: str):
+        """
+        Save the IVF model to a lance file.
+
+        Parameters
+        ----------
+
+        uri: str
+            The URI to save the model to.  The URI can be a local file path or a
+            cloud storage path.
+        """
+        with LanceFileWriter(
+            uri,
+            pa.schema(
+                [pa.field("centroids", self.centroids.type)],
+                metadata={b"distance_type": self.distance_type.encode()},
+            ),
+        ) as writer:
+            batch = pa.table([self.centroids], names=["centroids"])
+            writer.write_batch(batch)
+
+    @classmethod
+    def load(cls, uri: str):
+        """
+        Load an IVF model from a lance file.
+
+        Parameters
+        ----------
+
+        uri: str
+            The URI to load the model from.  The URI can be a local file path or a
+            cloud storage path.
+        """
+        reader = LanceFileReader(uri)
+        num_rows = reader.metadata().num_rows
+        metadata = reader.metadata().schema.metadata
+        distance_type = metadata[b"distance_type"].decode()
+        centroids = (
+            reader.read_all(batch_size=num_rows).to_table().column("centroids").chunk(0)
+        )
+        return cls(centroids, distance_type)
+
+
+class IndicesBuilder:
+    """
+    A class with helper functions for building indices on a dataset.
+
+    This methods in this class can break down the process of building indices into
+    smaller steps.  This can be useful for debugging and checkpointing when building
+    indices for extremely large datasets.
+
+    This class is intended for advanced users that need to create vector indices at
+    large scales.
+
+    The methods in this class are **experimental** and may change in future versions.
+
+    For datasets with 10s of millions or fewer rows it will likely be simpler to just
+    use the `create_index` method on the dataset object.
+    """
+
+    def __init__(self, dataset):
+        self.dataset = dataset
+
+    def train_ivf(
+        self,
+        column,
+        num_partitions=None,
+        *,
+        distance_type="l2",
+        accelerator: Optional[Union[str, "torch.Device"]] = None,
+        sample_rate: int = 256,
+        max_iters: int = 50,
+    ) -> IvfModel:
+        """
+        Train IVF centroids for the given vector column.
+
+        This will run k-means clustering on the given vector column to train the IVF
+        centroids. This is the first step in several vector indices.  The centroids
+        will be used to partition the vectors into different clusters.
+
+        K-means is an iterative algorithm that can be computationally expensive. The
+        accelerator argument can be used to offload the computation to a hardware
+        accelerator such as a GPU or TPU.
```

Comment:
Is there a progress bar for this? Should we mention the `LANCE_LOG` variable here?

Reply:
Yes, there is a tqdm progress bar if an accelerator is used.  I started to add one if an accelerator is not used but it ended up being a fair amount of code.  Will open a different PR.